cmd/cosign/cli/verify/verify.go

```diff
@@ -154,7 +154,7 @@ func (c *VerifyCommand) Exec(ctx context.Context, images []string) (err error) {
 		if err != nil {
 			return fmt.Errorf("getting Fulcio roots: %w", err)
 		}
-		co.IntermediateCerts, err = fulcio.GetIntermediates()
+		co.UntrustedIntermediateCerts, err = fulcio.GetIntermediates()
```

Comment:
Could we keep the same name, `IntermediateCerts`?

I would prefer we don't used "Untrusted" in the name. This term comes from openssl from what I've seen, and I personally don't think it's correct to only use intermediate certificates for chain building ("untrusted"). Intermediates should be trusted, or at least come from a trusted source.

Reply:
This one is confusing. These particular ones are coming from a trusted source, so I can agree to see it named the same. The ones found on the registry through CertChain are untrusted intermediates

Reply:
Yes, see `co.UntrustedIntermediateCerts = untrustedPool`. So, in that sense, I think changing the name of `CheckOpts` to `UntrustedIntermediateCerts` has been a success exactly because the trusted nature of those coming from `fulcio.GetIntermediates` is insufficient to make the field trusted.

(I’m also not sure what a “trusted intermediate” means. Is that a root of trust, i.e. a non-intermediate root? Or an intermediate that we still require to be correctly signed by a root of trust, i.e. something where we don’t actually require any trust?)

---

Actually, what I _really_ think is that the caller-provided policy should be immutable and should not be modified to hold single-instance data within `verify.go` at all. Then we might even more precisely differentiate between
- completely-untrusted signature-carried data
- externally Fulcio-provided data (acceptable if the connection to Fulcio is trusted)
- completely locally-provided 100%-trusted policy

and potentially make different decisions based on those details. That would be a different PR.

Reply:
For a core library function, I think we should not be opinionated if the certificates are trusted or untrusted. For example, we don't know where the root certificate came from either when the checkopts is populated. I don't see a reason to distinguish between untrusted and trusted in the CheckOpts. Internally, we can name variables differently based on their source.

Reply:
> I think we should not be opinionated if the certificates are trusted or untrusted.

I… don’t know what that would mean.

The verification code _must_ be certain about which data is a part of the root of trust that is, by necessity, assumed valid, and which data is part of the external attack surface that is, by default, suspect. At the very least these two categories are 100% distinct. There might be even more categories in between (e.g. the intermediate certificates might be more trusted than the underlying signatures, but they are still potentially a part of the attack surface).

(And then the API should help the caller of the API express its assumptions clearly, and to help expose any incorrect assumptions the caller might be making.

From that point of view, calling the certificates “untrusted” is _good_ for the caller, because it tells the caller the caller doesn’t need to make much effort at all to ensure validity of the data. “A trusted system is one whose failure would break a security policy”.)

---

I don’t care at all about the individual naming of the options. If you don’t want the field to be named `Untrusted`, and to instead have a `CheckOpts.{TrustedSigVerifier,TrustedRootCerts}`, sure, that might be even more explicit. If you think the distinction between trusted and untrusted data is sufficient to express in comments and doesn’t need to be in field names, that’s not my preference but I can live with that just fine.

But I read the above as suggesting that it doesn’t make a difference whether the data is trusted or not, and I just don’t know what to do with that. Probably I’m completely misunderstanding your point.